File: BAEKJOON/3190.py
# Dummy 도스 게임
# 벽 or 자기자신 몸과 부딪히면 게임 끝
# N*N 몇몇 칸 사과 AND  보드 상하좌우 끝에 벽
# 뱀 0,0 위치 
# 몸길이 늘려 다음칸으로 -> 사과 있으면 꼬리 그대로// 없으면 꼬리 당겨서 길이 유지


# 보드 크기 N, 사과 갯수 K, 방향 변환 횟수 L
# 정수 X, 문자 C -> X초 뒤에 왼쪽 or 오른쪽으로 90도 회전 

# 문제는 어떻게 하느냐.. q를 이용??
# 오른쪽 [[0,1],[1,0],[0,-1],[1,0]]
import sys
from collections import deque
input = sys.stdin.readline
N = int(input())
matrix = [[0]*N for _ in range(N)] 
for _ in range(int(input())):
    a,b = map(int,input().split())
    matrix[a-1][b-1] = 1
rotate = {}
for _ in range(int(input())):
    time,cmd = input().split()
    rotate[int(time)] = cmd
xy = [[0,1],[1,0],[0,-1],[-1,0]]
time = 1
snake = deque([[0,0,0]]) # row,col,direction
matrix[0][0] = - 1
# 몸통만 저장해두고.
# 사과 없으면 꼬리pop() & direction 지정해서 몸통 append

while True:
    head,d = snake[-1][:2],snake[-1][2] # row,col만
    head[0] += xy[d][0]
    head[1] += xy[d][1]
    if 0<= head[0] <N and 0<=head[1] <N and matrix[head[0]][head[1]] != -1: # 몸통 & 범위 밖아님
        if matrix[head[0]][head[1]]: # 사과있음
            pass
        else:
            x,y,_ = snake.popleft()
            matrix[x][y] = 0
        matrix[head[0]][head[1]] = -1
        if time in rotate:
            if rotate[time] == 'L': # 왼쪽 회전
                d = (d-1) % 4
            else: # 오른쪽 회전
                d = (d+1) % 4
        time += 1
        snake.append([head[0],head[1],d])
    else:
        print(time)
        break

File: BAEKJOON/3190.java
import java.util.*;
import java.io.*;
public class Main {
	public static void main(String[] args) throws  IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		HashMap<Integer, String> h = new HashMap<>();
		int[][] arr = new int[N][N];
		int t = Integer.parseInt(br.readLine());
		for(int i = 0; i < t; i ++) {
			String[] tt = br.readLine().split(" ");
			arr[Integer.parseInt(tt[0]) - 1][Integer.parseInt(tt[1]) - 1] = 1;
		}
		t = Integer.parseInt(br.readLine());
		for(int i = 0; i < t; i ++) {
			String[] tt = br.readLine().split(" ");
			h.put(Integer.parseInt(tt[0]), tt[1]);
		}
		System.out.println(sol(h,arr));
	}
		
	
	
	public static int sol(HashMap<Integer, String> h, int[][] arr ) {
		int[][] d = {{0,1}, {1,0},{0,-1},{-1,0}};
		List<int[]> snake = new ArrayList<>();
		arr[0][0] = 2;
		snake.add(new int[] {0,0});
		int direction = 0;
		int time = 0;
		while(true) {
			time ++;
			int[] next = {snake.get(0)[0] + d[direction][0],snake.get(0)[1] + d[direction][1] };
			if(!(0<= next[0] && next[0] < arr.length && next[1] < arr.length && next[1] >= 0)) return time;
			if(arr[next[0]][next[1]] == 2) return time;
			if (arr[next[0]][next[1]] == 1) {
				snake.add(0,next); // 사과 먹으면 그냥 몸 늘어감
			}
			else {
				arr[snake.get(snake.size() -1)[0]][snake.get(snake.size() -1)[1]] = 0; // 제일 끝 지워줌
				move(snake,next);
			}
			arr[next[0]][next[1]] = 2; // 머리 늘어감
			if (h.containsKey(time)) {
				direction = rotate(direction, h.get(time));
			}

		}
		
	}
	
	
	public static int rotate(int direction, String d) {
		if (d.equals("L")) {
			direction = direction - 1 > -1 ? direction - 1 : 3 ;
		}
		else {
			direction = direction + 1 < 4 ? direction + 1 : 0 ;
		}
		return direction;
	}
	
	public static void move(List<int[]> snake, int[] next) {
		for(int i = 0; i < snake.size(); i++) {
			int[] t = snake.get(i);
			snake.set(i, next);
			next = t;
		}
	}
}

